src/test/metrics_api_test.cc

```diff
@@ -236,30 +344,239 @@ TEST_F(MetricsApiTest, TestGaugeEndToEnd)
+  // The first description will be taken/kept if adding a duplicate
+  /// metric family name, even with a different description
+  ASSERT_EQ(NumMetricMatches(server_, description2), 0);
+
+  // Delete one of the metric family references
+  // Specificailly, family1, because family2 is bound to metric2
+  FAIL_TEST_IF_ERR(TRITONSERVER_MetricFamilyDelete(family1), "delete family1");
+
+  // Assert custom metric/family remains when family2 still references it
+  ASSERT_EQ(NumMetricMatches(server_, description1), 1);
+
+  // Assert custom metric/family unregistered after last reference deleted
+  FAIL_TEST_IF_ERR(TRITONSERVER_MetricDelete(metric2), "delete metric2");
+  FAIL_TEST_IF_ERR(TRITONSERVER_MetricFamilyDelete(family2), "delete family2");
+  ASSERT_EQ(NumMetricMatches(server_, description1), 0);
+  ASSERT_EQ(NumMetricMatches(server_, description2), 0);
+}
+
+// Test that adding a duplicate metric family will reuse the original
+// and not add another entry to registry
+TEST_F(MetricsApiTest, TestDupeMetricFamily)
+{
+  // Create metric family
+  TRITONSERVER_MetricFamily* family1 = nullptr;
+  TRITONSERVER_MetricKind kind = TRITONSERVER_METRIC_KIND_COUNTER;
+  const char* name = "dupe_metric_family_test";
+  const char* description = "dupe metric family description";
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricFamilyNew(&family1, kind, name, description),
+      "Creating new metric family1");
+
+  // Create duplicate metric family
+  TRITONSERVER_MetricFamily* family2 = nullptr;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricFamilyNew(&family2, kind, name, description),
+      "Creating new metric family2");
+
+  // Assert MetricFamily is not reported until metrics are added to them
+  ASSERT_EQ(NumMetricMatches(server_, description), 0);
+
+  // Create unique metrics for each family object. Both family objects
+  // will refer to the same prometheus family in the registry, so both
+  // metrics should be displayed under the family.
+  const char* metric_key = "custom_metric_key";
+  std::vector<const TRITONSERVER_Parameter*> labels1;
+  labels1.emplace_back(TRITONSERVER_ParameterNew(
+      metric_key, TRITONSERVER_PARAMETER_STRING, "label1"));
+  TRITONSERVER_Metric* metric1 = nullptr;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricNew(&metric1, family1, labels1.data(), labels1.size()),
+      "Creating new metric1");
+  for (const auto label : labels1) {
+    TRITONSERVER_ParameterDelete(const_cast<TRITONSERVER_Parameter*>(label));
+  }
+
+  std::vector<const TRITONSERVER_Parameter*> labels2;
+  labels2.emplace_back(TRITONSERVER_ParameterNew(
+      metric_key, TRITONSERVER_PARAMETER_STRING, "label2"));
+  TRITONSERVER_Metric* metric2 = nullptr;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricNew(&metric2, family2, labels2.data(), labels2.size()),
+      "Creating new metric2");
+  for (const auto label : labels2) {
+    TRITONSERVER_ParameterDelete(const_cast<TRITONSERVER_Parameter*>(label));
```

Comment:
I think it would be good to update the backend API documentation to mention that `TRITONSERVER_MetricNew` does not take the ownership of labels and the caller must delete them.